File: macos-app/AWMKit/Sources/Localization/UiErrorMapper.swift
import Foundation

struct UiMappedMessage {
    let resultTitle: String
    let userReason: String
    let nextAction: String
    let diagnosticCode: String
    let diagnosticDetail: String
    let rawError: String
    let techFields: [String: String]

    var userDetail: String {
        if nextAction.isEmpty {
            return userReason
        }
        if userReason.isEmpty {
            return nextAction
        }
        return "\(userReason)\n\(nextAction)"
    }
}

enum UiErrorMapper {
    static func map(title: String, detail: String, isSuccess: Bool) -> UiMappedMessage {
        let normalized = detail.trimmingCharacters(in: .whitespacesAndNewlines)
        if isSuccess {
            return UiMappedMessage(
                resultTitle: title,
                userReason: normalized,
                nextAction: "",
                diagnosticCode: "",
                diagnosticDetail: "",
                rawError: "",
                techFields: [:]
            )
        }

        let lowered = normalized.lowercased()
        var techFields: [String: String] = [:]
        if let route = extractToken("route", from: normalized) {
            techFields["route"] = route
        }
        if let status = extractToken("status", from: normalized) {
            techFields["status"] = status
        }

        if lowered.contains("single_fallback")
            || lowered.contains("unverified")
            || lowered.contains("invalid_hmac")
            || lowered.contains("route=")
            || lowered.contains("status=") {
            return UiMappedMessage(
                resultTitle: title,
                userReason: Localizer.pick("操作失败。", "The operation failed."),
                nextAction: Localizer.pick("下一步：打开诊断并重试。", "Next: Turn on diagnostics and retry."),
                diagnosticCode: "diag.internal_state",
                diagnosticDetail: normalized,
                rawError: normalized,
                techFields: techFields
            )
        }

        if lowered.contains("no such file") || lowered.contains("not found") || lowered.contains("path") {
            return UiMappedMessage(
                resultTitle: title,
                userReason: Localizer.pick("操作失败。", "The operation failed."),
                nextAction: Localizer.pick("下一步：检查输入路径后重试。", "Next: Check the input path and retry."),
                diagnosticCode: "diag.path",
                diagnosticDetail: normalized,
                rawError: normalized,
                techFields: techFields
            )
        }

        return UiMappedMessage(
            resultTitle: title,
            userReason: Localizer.pick("操作失败。", "The operation failed."),
            nextAction: Localizer.pick("下一步：请重试；若再次失败，请打开诊断。", "Next: Retry, and turn on diagnostics if it fails again."),
            diagnosticCode: "diag.generic",
            diagnosticDetail: normalized,
            rawError: normalized,
            techFields: techFields
        )
    }

    private static func extractToken(_ key: String, from source: String) -> String? {
        let pattern = "\\b\\(key)=([^\\s,;]+)"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return nil
        }
        let range = NSRange(source.startIndex..<source.endIndex, in: source)
        guard let match = regex.firstMatch(in: source, options: [], range: range),
              let tokenRange = Range(match.range(at: 1), in: source) else {
            return nil
        }
        return String(source[tokenRange])
    }
}

File: winui-app/AWMKit/Localization/UiErrorMapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace AWMKit.Localization;

public readonly record struct UiMappedMessage(
    string ResultTitle,
    string UserReason,
    string NextAction,
    string DiagnosticCode,
    string DiagnosticDetail,
    string RawError,
    string TechFields)
{
    public string UserDetail
    {
        get
        {
            if (string.IsNullOrWhiteSpace(NextAction))
            {
                return UserReason;
            }

            if (string.IsNullOrWhiteSpace(UserReason))
            {
                return NextAction;
            }

            return $"{UserReason}\n{NextAction}";
        }
    }
}

public static class UiErrorMapper
{
    private static readonly Regex TechTokenRegex = new(@"\b(route|status)=[^\s,;]+", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    public static UiMappedMessage Map(string title, string detail, bool isSuccess)
    {
        var normalized = detail?.Trim() ?? string.Empty;
        if (isSuccess)
        {
            return new UiMappedMessage(title, normalized, string.Empty, string.Empty, string.Empty, string.Empty, string.Empty);
        }

        var techFields = string.Join(", ", TechTokenRegex.Matches(normalized).Select(m => m.Value).Distinct(StringComparer.OrdinalIgnoreCase));
        var hasInternalField = normalized.Contains("single_fallback", StringComparison.OrdinalIgnoreCase)
            || normalized.Contains("UNVERIFIED", StringComparison.OrdinalIgnoreCase)
            || normalized.Contains("invalid_hmac", StringComparison.OrdinalIgnoreCase)
            || normalized.Contains("status=", StringComparison.OrdinalIgnoreCase)
            || normalized.Contains("route=", StringComparison.OrdinalIgnoreCase);

        if (hasInternalField)
        {
            return new UiMappedMessage(
                title,
                AppStrings.Get("ui.error.processing_failed"),
                AppStrings.Get("ui.error.next.open_diagnostics"),
                "diag.internal_state",
                normalized,
                normalized,
                techFields);
        }

        if (normalized.Contains("No such file", StringComparison.OrdinalIgnoreCase)
            || normalized.Contains("not found", StringComparison.OrdinalIgnoreCase)
            || normalized.Contains("path", StringComparison.OrdinalIgnoreCase))
        {
            return new UiMappedMessage(
                title,
                AppStrings.Get("ui.error.processing_failed"),
                AppStrings.Get("ui.error.next.check_input_path"),
                "diag.path",
                normalized,
                normalized,
                techFields);
        }

        return new UiMappedMessage(
            title,
            AppStrings.Get("ui.error.processing_failed"),
            AppStrings.Get("ui.error.next.retry"),
            "diag.generic",
            normalized,
            normalized,
            techFields);
    }
}
